tools/tpm2_convert_key.c

```diff
@@ -0,0 +1,414 @@
+#include "tpm2_util.h"
+
+/*
+ * This program takes a key file as created by `tpm2_akparse` as input and
+ * converts it into OpenSSL PEM or DER format public keys.
+ *
+ * This is currently limited to RSA public keys but might be extended to other
+ * key types in the future.
+ */
+
+// holds all the custom command line parameters for our purposes
+typedef struct tpm_convert_key_ctx tpm_convert_key_ctx;
+struct tpm_convert_key_ctx {
+    char *key_input_file_path;
+    char *key_output_file_path;
+    char *key_output_format;
+};
+
+// holds a single key blob as found in the input file
+struct key_blob {
+    UINT16 length;
+    UINT8  *data;
+};
+
+#define MAX_KEY_BLOBS 8
+
+// holds the complete information as found in the input file
+struct key_data {
+    UINT16 alg_type;
+    size_t num_blobs;
+    struct key_blob blobs[MAX_KEY_BLOBS];
+};
+
+typedef enum {
+    FMT_PEM,
+    FMT_DER,
+    FMT_INVAL
+} convert_format;
+
+// reads just the algorithm type identifier from the given file and returns it
+// in the correct endianess. error handling included.
+static bool read_alg_type(FILE *f, UINT16 *alg_type) {
+
+    if (fread(alg_type, 2, 1, f) != 1) {
+        LOG_ERR("Could not read algtype from input file: \"%s\".",
+            (feof(f) ? "Premature end of file" : strerror(errno))
+        );
+
+        return false;
+    }
+
+    // convert byte order
+    *alg_type = tpm2_util_ntoh_16(*alg_type);
+
+    return true;
+}
+
+// frees any dynamically allocated data in the given key_data struct
+static void free_keys(struct key_data *data) {
+    size_t i;
+    struct key_blob *blob;
+
+    for (i = 0; i < data->num_blobs; i++) {
+        blob = &(data->blobs[i]);
+        free(blob->data);
+        blob->data = NULL;
+    }
+}
+// parse all found blob parts from the given file and add them to the key_data
+// struct. error handling included.
+// on true return the caller is responsible for freeing data
+static bool read_blobs(FILE *f, struct key_data *data) {
+    const size_t MAX_LENGTH = 8192;
+    struct key_blob blob;
+    bool ret = false;
+
+    // read in a variable number of blobs until EOF or error occurs
+    while (true) {
+
+        if (fread(&blob.length, 2, 1, f) != 1) {
+            if (!feof(f)) {
+                LOG_ERR("Could not read key blob length from input file: \"%s\".",
+                    strerror(errno)
+                );
+
+                goto out;
+            }
+
+            ret = true;
+            break;
+        }
+
+        // correct byte order
+        blob.length = tpm2_util_ntoh_16(blob.length);
+
+        if (blob.length > MAX_LENGTH) {
+            LOG_ERR("Excess length key blob of %d bytes encountered",
+                blob.length
+            );
+
+            goto out;
+        }
+        else if (data->num_blobs == MAX_KEY_BLOBS) {
+            LOG_ERR("Maximum number of key blobs %d exceeded.", MAX_KEY_BLOBS);
+            goto out;
+        }
+
+        blob.data = malloc(blob.length);
```

Comment:
bolb.data possible null pointer de-reference when malloc fails.

Reply:
is handled now